File: 2-NORMAL/Python/9-word-ladder-challenge/proposed-solution/word_ladder.py
from collections import deque

def word_ladder(begin_word, end_word, word_list):
    word_set = set(word_list)
    if end_word not in word_set:
        return []

    queue = deque([(begin_word, [begin_word])])
    visited = set([begin_word])

    while queue:
        current_word, path = queue.popleft()
        if current_word == end_word:
            return path

        for i in range(len(current_word)):
            for c in 'abcdefghijklmnopqrstuvwxyz':
                next_word = current_word[:i] + c + current_word[i+1:]
                if next_word in word_set and next_word not in visited:
                    visited.add(next_word)
                    queue.append((next_word, path + [next_word]))

    return []

File: 2-NORMAL/JavaScript/9-word-ladder-challenge/proposed-solution/wordLadder.js
class WordLadder {
  static wordLadder(beginWord, endWord, wordList) {
    const wordSet = new Set(wordList);
    if (!wordSet.has(endWord)) {
      return [];
    }

    const queue = [[beginWord, [beginWord]]];
    const visited = new Set([beginWord]);

    while (queue.length > 0) {
      const [currentWord, path] = queue.shift();
      if (currentWord === endWord) {
        return path;
      }

      for (let i = 0; i < currentWord.length; i++) {
        for (let c = 97; c <= 122; c++) { // 'a' to 'z'
          const char = String.fromCharCode(c);
          const nextWord = currentWord.slice(0, i) + char + currentWord.slice(i + 1);
          if (wordSet.has(nextWord) && !visited.has(nextWord)) {
            visited.add(nextWord);
            queue.push([nextWord, [...path, nextWord]]);
          }
        }
      }
    }

    return [];
  }
}

export default WordLadder;
